### PR TITLE
feat(api): Refactor API server to introduce support for Docker image registries and s3 bucket support

### PR DESCRIPTION
**Pull request by deadlycoconuts**
## Note 🚨
~~This PR should not be merged without the changes in https://github.com/caraml-dev/mlp/pull/116 being merged, published and imported as dependencies first. This PR is currently only using a branch of a fork (the source of the quoted PR) of the MLP repository.~~ The dependent PR has been merged.

# Description
In order to provide support for using image registries that use [Docker registry credentials](https://docs.docker.com/reference/cli/docker/login/#configure-the-credential-store) as well as [AWS S3-based blob storage services](https://docs.aws.amazon.com/code-library/latest/ug/go_2_s3_code_examples.html), this PR refactors the API server to support these additional image registry and bob storage options. More concretely, these are the following changes made:

- Set up the workflow needed to allow platform maintainers to configure the image registry that the API server will push images to (Docker or Google Cloud/Artifact Registry) as well as the blob storage service that it should read model artifacts from/write files to (S3-based store/Google Cloud Storage)
- Allow the API server to access a configured Docker registry to check if an image is available
- Allow the API server to check and hash model dependencies in a configured S3-based store
- Allow Kaniko jobs spun up by the API server to use load model artifacts from a configured S3-based store when building model images
- Allow Kaniko jobs spun up by the API server to push images to the configured Docker registry

# Modifications
- `api/cmd/api/setup.go` - Make the initialisation of the image builder set up the artifact service type and docker registry correctly depending on the one set up
- `api/config/config.go` - Introduce new configs for platform maintainers to specify the `KanikoPushRegistryType` and the `KanikoDockerCredentialSecretName`
- `api/pkg/imagebuilder/imagebuilder.go` - Make changes to the image builder to configure the Kaniko job spec correctly depending on the selected registry type and blob storage type
- `python/batch-predictor/docker/app.Dockerfile` - Add steps to the batch predictor docker image to authenticate and pull model artifacts correctly depending on the configured blob storage type 
- `python/batch-predictor/docker/base.Dockerfile` - Add steps to the base batch predictor image to install the AWS CLI
- `python/pyfunc-server/docker/Dockerfile` - Add steps to the pyfunc server docker image to authenticate and pull model artifacts correctly depending on the configured blob storage type
- `python/pyfunc-server/docker/base.Dockerfile` - Add steps to the base pyfunc server image to install the AWS CLI

# Tests

# Checklist
- [x] Added PR label
- [x] Added unit test, integration, and/or e2e tests
- [x] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduces API changes

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: api/cmd/api/setup.go
Comment (deadlycoconuts) [inline api/cmd/api/setup.go:108]: The need to manually initialise a GCS client is gone with the refactoring done in https://github.com/caraml-dev/mlp/pull/116 (the imported `NewGcsArtifactClient` will do it instead).

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [inline api/pkg/imagebuilder/imagebuilder.go:173]: The conversion of `condaEnv` to bytes is unnecessary since it's already of the type `[]byte`.

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [inline api/pkg/imagebuilder/imagebuilder.go:623]: This argument gets passed to the batch predictor/pyfunc server image, and is used to determine the correct CLI (GCS or AWS) to use to pull the model artifacts from.

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [inline api/pkg/imagebuilder/imagebuilder.go:637]: For now I'm just reading this values from the env vars in the environment where the API server in run in. Does anyone have an opinion if we should move it into the config file of the API server? 😅

File: api/go.mod
Comment (deadlycoconuts) [inline api/go.mod:277]: I'm temporarily pointing this to my branch but I'll update this once the MLP PR has been approved and merged.

File: api/config/config.go
Comment (deadlycoconuts) [inline api/config/config.go:214]: I removed the field `ArtifactServiceType` here because it's essentially the same as `MlflowConfig.ArtifactServiceType`.

File: api/cmd/api/setup.go
Comment (bthari) [inline api/cmd/api/setup.go:122]: very nit: but the err on this line belong to this [if else](https://github.com/caraml-dev/merlin/blob/a63f15f06e2801c9392d965586846307a2d720bf/api/cmd/api/setup.go#L107-L115), so maybe can put it after this

File: api/pkg/imagebuilder/imagebuilder.go
Comment (bthari) [inline api/pkg/imagebuilder/imagebuilder.go:650]: does this mean to be "gcs" or does [this line](https://github.com/caraml-dev/merlin/blob/a63f15f06e2801c9392d965586846307a2d720bf/api/cmd/api/setup.go#L107) should be "gcr", I supposed these two should be the same?

File: api/pkg/imagebuilder/imagebuilder.go
Comment (tiopramayudi) [inline api/pkg/imagebuilder/imagebuilder.go:677]: can we have interface for dockerRegistry and pass it to image builder? so we don't need to have condition here

File: python/batch-predictor/docker/app.Dockerfile
Comment (tiopramayudi) [inline python/batch-predictor/docker/app.Dockerfile:27]: would it better to separate dockerfile based on cloud storage or cloud provider?

File: api/cmd/api/setup.go
Comment (deadlycoconuts) [reply api/cmd/api/setup.go:122]: Oh do you mean like move this block
```go
if err != nil {
	log.Panicf("%s,failed initializing mlflow artifact service", err.Error())
}
```
to the part just after this
```go
if cfg.MlflowConfig.ArtifactServiceType == "gcs" {
	artifactService, err = artifact.NewGcsArtifactClient()
} else if cfg.MlflowConfig.ArtifactServiceType == "s3" {
	artifactService, err = artifact.NewS3ArtifactClient()
} else if cfg.MlflowConfig.ArtifactServiceType == "nop" {
	artifactService = artifact.NewNopArtifactClient()
} else {
	log.Panicf("invalid artifact service type %s", cfg.MlflowConfig.ArtifactServiceType)
}
```
so the final result looks like this?
```go
if cfg.MlflowConfig.ArtifactServiceType == "gcs" {
	artifactService, err = artifact.NewGcsArtifactClient()
} else if cfg.MlflowConfig.ArtifactServiceType == "s3" {
	artifactService, err = artifact.NewS3ArtifactClient()
} else if cfg.MlflowConfig.ArtifactServiceType == "nop" {
	artifactService = artifact.NewNopArtifactClient()
} else {
	log.Panicf("invalid artifact service type %s", cfg.MlflowConfig.ArtifactServiceType)
}
if err != nil {
	log.Panicf("%s,failed initializing mlflow artifact service", err.Error())
}
```

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:650]: 😱 nice catch! It's the other one that should be `gcr` hahaha 😅 Yes both of them should be the same!

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:649]: Ah do you mean like refactor this block into a separate helper function? Yep that'll make it cleaner :)

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:677]: Ah do you mean an interface for the `imageBuilder` and then have 2 implementations of it, i.e. `dockerImageBuilder` and `gcrImageBuilder`?

File: python/batch-predictor/docker/app.Dockerfile
Comment (deadlycoconuts) [reply python/batch-predictor/docker/app.Dockerfile:27]: Yeah I was thinking of that too, but that'll require some additional rework of our CICD pipelines + publishing new images.. but yeah I can do that.

File: api/cmd/api/setup.go
Comment (bthari) [reply api/cmd/api/setup.go:122]: Yess, I'm thinking of moving it like that, because the `err` belongs to that if else codeblock

File: api/pkg/imagebuilder/imagebuilder.go
Comment (tiopramayudi) [reply api/pkg/imagebuilder/imagebuilder.go:677]: yes something like that, but this is nit. We can revisit this later

File: api/pkg/imagebuilder/imagebuilder.go
Comment (tiopramayudi) [reply api/pkg/imagebuilder/imagebuilder.go:649]: yes

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:637]: Hmm what do you mean by that?

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:649]: Hmm, I've currently refactored the entire code chunk into tinier helper functions but they are currently grouped by the types of configs/specs that are being modified (instead of the type of push registry):
```go
	// Configure additional authentication requirements for specific image registries
	kanikoArgs = c.configureKanikoArgsForKanikoPushRegistry(kanikoArgs)
	volumes, volumeMounts = c.configureVolumesAndVolumeMountsForKanikoPushRegistry(volumes, volumeMounts)
	envVar = c.configureEnvVarsForKanikoPushRegistry(envVar)
```

File: api/pkg/imagebuilder/imagebuilder.go
Comment (tiopramayudi) [reply api/pkg/imagebuilder/imagebuilder.go:637]: I mean we can move it into config file and add it as kaniko default argument

File: api/cmd/api/setup.go
Comment (deadlycoconuts) [reply api/cmd/api/setup.go]: Got it, it should be updated now! Thanks!

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go]: Okay I've removed this entire code chunk, assuming that the user of Merlin will be adding these additional values to the `KanikoAdditionalArgs` [field](https://github.com/caraml-dev/merlin/pull/605/files#diff-7f88f1945fb732d6ac18c0a4b3adfd6bec6f337538050f52d4a52db2e1d65b47R223) of the config file. I've added some additional doc strings [here](https://github.com/caraml-dev/merlin/pull/605/files#diff-7f88f1945fb732d6ac18c0a4b3adfd6bec6f337538050f52d4a52db2e1d65b47R458) to inform them that they'd need to ensure that the kaniko job runner has the necessary credentials to download model artifacts from the blob storage configured.

File: python/batch-predictor/docker/app.Dockerfile
Comment (deadlycoconuts) [reply python/batch-predictor/docker/app.Dockerfile:27]: For this part, we've decided to work on this another time as an enhancement as it requires more extensive reworking of the existing CICD pipelines as well as Merlin config files to work. I'll create another card in our backlog to keep track of this.

**Comment by deadlycoconuts** [comment]
Thanks a lot for the comments @tiopramayudi and @bthari ! I should've covered and addressed all of the outstanding comments. If all looks good feel free to approve it again otherwise let me know what additional changes we should include and I'll get back to it. Thanks once again!

**Review by bthari — APPROVED** [review]
Thank you for the changes for this and Turing, and for going through all the feedback! Left two tiny comments, but the rest is LGTM, so feel free to merge this once you've addressed it

File: .github/workflows/merlin.yml
Comment (deadlycoconuts) [reply .github/workflows/merlin.yml:140]: Ah it's because the test coverage scan is inaccurate. Sometimes even without making any code changes, the code coverage report reports a massive % drop and fails the CICD job :/

File: api/pkg/imagebuilder/imagebuilder.go
Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:746]: Good catch! I'll fix this!

Comment (deadlycoconuts) [reply api/pkg/imagebuilder/imagebuilder.go:746]: I'll also refactor the `"gcs"`, "`gcr"`, and `"docker"` strings into constants like Tio suggested in the other PR.

File: python/sdk/merlin/util.py
Comment (deadlycoconuts) [inline python/sdk/merlin/util.py:119]: @tiopramayudi I've added this new block in order to download files directly using the AWS boto3 client from an s3 blob storage to his/her local machine. The user would still need to ensure he/she is has the necessary authentication/connection configs ready (as env vars), just like how it's expected when he/she uses GCS, in order for this to work.

**Comment by deadlycoconuts** [comment]
Thanks for the re-review @bthari! I'll keep this PR open a little while longer if Tio wants to look at the part I added to the Merlin SDK to allow users to download the Merlin model files from an s3-configured Mlflow blob store. Otherwise I'll merge this by tomorrow! :D

**Review by tiopramayudi — APPROVED** [review]

**Pull request merged**
